larpandoracontent/LArControlFlow/NeutrinoIdTool.cc

```diff
@@ -10,30 +10,464 @@
+
+void NeutrinoIdTool::SelectAllPfos(const SliceHypotheses &hypotheses, PfoList &selectedPfos) const
+{
+    for (const PfoList &pfos : hypotheses)
+        this->SelectPfos(pfos, selectedPfos);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+    
+void NeutrinoIdTool::SelectPfosByProbability(const SliceHypotheses &nuSliceHypotheses, const SliceHypotheses &crSliceHypotheses, const SliceFeaturesVector &sliceFeaturesVector, PfoList &selectedPfos) const
+{
+    // Calculate the probability of each slice that passes the minimum probability cut
+    std::vector<UintFloatPair> sliceIndexProbabilityPairs;
+    for (unsigned int sliceIndex = 0, nSlices = nuSliceHypotheses.size(); sliceIndex < nSlices; ++sliceIndex)
+    {
+        const float nuProbability(sliceFeaturesVector.at(sliceIndex).GetNeutrinoProbability(m_supportVectorMachine));
+
+        if (nuProbability < m_minProbability)
+        {
+            this->SelectPfos(crSliceHypotheses.at(sliceIndex), selectedPfos);
+            continue;
+        }
+
+        sliceIndexProbabilityPairs.push_back(UintFloatPair(sliceIndex, nuProbability));
+    }
+
+    // Sort the slices by probability
+    std::sort(sliceIndexProbabilityPairs.begin(), sliceIndexProbabilityPairs.end(), [] (const UintFloatPair &a, const UintFloatPair &b) 
+    {
+        return (a.second > b.second);
+    });
+
+    // Select the first m_maxNeutrinos as neutrinos, and the rest as cosmic
+    unsigned int nNuSlices(0);
+    for (const UintFloatPair &slice : sliceIndexProbabilityPairs)
+    {
+        if (nNuSlices < m_maxNeutrinos)
+        {
+            this->SelectPfos(nuSliceHypotheses.at(slice.first), selectedPfos);
+            nNuSlices++;
+            continue;
+        }
+        
+        this->SelectPfos(crSliceHypotheses.at(slice.first), selectedPfos);
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void NeutrinoIdTool::SelectPfos(const PfoList &pfos, PfoList &selectedPfos) const
+{
+    selectedPfos.insert(selectedPfos.end(), pfos.begin(), pfos.end());
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+// TODO think about how to make this function cleaner when features are more established
+NeutrinoIdTool::SliceFeatures::SliceFeatures(const PfoList &nuPfos, const PfoList &crPfos, const NeutrinoIdTool *const pTool) : 
+    m_isAvailable(false), m_pTool(pTool)
+{
+    const ParticleFlowObject *const pNeutrino(this->GetNeutrino(nuPfos));
+    const CartesianVector nuVertex(LArPfoHelper::GetVertex(pNeutrino)->GetPosition());
```

Comment:
Can this function raise an exception?

Can this be a const reference?

Reply:
You are right, good point, it can raise an exception. In fact, I want to bail out if the calculation of any feature isn't possible in any way, this could be due to numerical fails (e.g. all of the zero checks), or if *anything* raises and exception. I have now wrapped everything in a try block to ensure safety against this.